File: ios/Sources/CapacitorVoiceRecorderPlugin/PcmToWavConverter.swift
import AVFoundation

class PcmToWavConverter {
    private let sampleRate: Double
    private let channels: AVAudioChannelCount
    private let bitsPerSample: Int
    
    init(_ sampleRate: Double, _ channels: AVAudioChannelCount, _ bitsPerSample: Int) {
        self.sampleRate = sampleRate
        self.channels = channels
        self.bitsPerSample = bitsPerSample
    }
    
    /// Converts PCM data to WAV format.
    /// - Parameter pcmData: The raw PCM data.
    /// - Returns: A `Data` object containing the WAV formatted data, or `nil` if the conversion fails.
    func convertToWav(pcmData: Data) -> Data {
        let byteRate = Int(sampleRate) * Int(channels) * (bitsPerSample / 8)
        let blockAlign = Int(channels) * (bitsPerSample / 8)
        let dataSize = UInt32(pcmData.count)
        let totalSize = 44 + Int(dataSize)
        
        // Helper function to convert string to data
        func stringToData(_ string: String) -> [UInt8] {
            return Array(string.utf8)
        }
        
        var header = Data()
        
        // RIFF chunk descriptor
        header.append(contentsOf: stringToData("RIFF"))
        header.append(contentsOf: UInt32(totalSize - 8).toLittleEndian())
        header.append(contentsOf: stringToData("WAVE"))
        
        // fmt subchunk
        header.append(contentsOf: stringToData("fmt "))
        header.append(contentsOf: UInt32(16).toLittleEndian())
        header.append(contentsOf: [0x01, 0x00]) // AudioFormat (PCM = 1)
        header.append(contentsOf: UInt16(channels).toLittleEndian())
        header.append(contentsOf: UInt32(sampleRate).toLittleEndian())
        header.append(contentsOf: UInt32(byteRate).toLittleEndian())
        header.append(contentsOf: UInt16(blockAlign).toLittleEndian())
        header.append(contentsOf: UInt16(bitsPerSample).toLittleEndian())
        
        // data subchunk
        header.append(contentsOf: stringToData("data"))
        header.append(contentsOf: UInt32(dataSize).toLittleEndian())
        
        // Combine Header and PCM Data
        var wavData = header
        wavData.append(pcmData)
        
        return wavData
    }
}

extension FixedWidthInteger {
    /// Converts the integer to an array of bytes in little-endian order.
    func toLittleEndian() -> [UInt8] {
        withUnsafeBytes(of: self.littleEndian) { Array($0) }
    }
}

File: android/src/main/java/com/lgicc/capacitor/voice_recorder/PcmToWavConverter.java
package com.lgicc.capacitor.voice_recorder;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class PcmToWavConverter {

    public static byte[] convertPCMToWAV(byte[] pcmData, int sampleRate, int numChannels) throws IOException {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();

        // WAV header
        int byteRate = sampleRate * numChannels * 2; // 16-bit = 2 bytes per sample
        int blockAlign = numChannels * 2;

        // RIFF Header
        byteArrayOutputStream.write("RIFF".getBytes());
        int chunkSize = 36 + pcmData.length; // 36 for the WAV header size and the PCM data size
        byteArrayOutputStream.write(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(chunkSize).array());

        // WAVE header
        byteArrayOutputStream.write("WAVE".getBytes());

        // fmt Chunk
        byteArrayOutputStream.write("fmt ".getBytes());
        byteArrayOutputStream.write(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(16).array()); // Subchunk1Size
        byteArrayOutputStream.write(ByteBuffer.allocate(2).order(ByteOrder.LITTLE_ENDIAN).putShort((short) 1).array()); // Audio format (1 = PCM)
        byteArrayOutputStream.write(ByteBuffer.allocate(2).order(ByteOrder.LITTLE_ENDIAN).putShort((short) numChannels).array()); // NumChannels
        byteArrayOutputStream.write(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(sampleRate).array()); // SampleRate
        byteArrayOutputStream.write(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(byteRate).array()); // ByteRate
        byteArrayOutputStream.write(ByteBuffer.allocate(2).order(ByteOrder.LITTLE_ENDIAN).putShort((short) blockAlign).array()); // BlockAlign
        byteArrayOutputStream.write(ByteBuffer.allocate(2).order(ByteOrder.LITTLE_ENDIAN).putShort((short) 16).array()); // BitsPerSample

        // data Chunk
        byteArrayOutputStream.write("data".getBytes());
        byteArrayOutputStream.write(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(pcmData.length).array()); // Subchunk2Size

        // Write the PCM data
        byteArrayOutputStream.write(pcmData);

        return byteArrayOutputStream.toByteArray();
    }
}
